### DISCUSSION TITLE
"Bad state: This database has already been closed" error in unit-tests

### DISCUSSION THREAD

**Discussion by rubiktubik** [simolus3/drift#1395] [Q&A] [answered]
Hello, I followed the [guide](https://moor.simonbinder.eu/docs/testing/) to setup my unit-test with moor.

When i have one unit-test everything runs fine, but with one additional test i get the

`Bad state: This database has already been closed`

error.

My database-class:
```dart
class MyDatabase extends _$MyDatabase {
  // we tell the database where to store the data with this constructor
  MyDatabase(QueryExecutor e) : super(e);
...
```

My test setup:

```dart
late MyDatabase _database;
late DbDataProvider _dataProvider;

  setUp(() {
    _database = MyDatabase(VmDatabase.memory());
    _dataProvider = DbDataProvider(_database);
  });

  tearDown(() async {
    await _database.close();
  });
```

My tests:

```dart
test('can create workout in db', () async {
    // Arrange
    Execution execution = Execution(currentRepetitons: 9, started: true);
    Execution execution2 = Execution(currentRepetitons: 8, started: true);
    Execution execution3 = Execution(currentRepetitons: 7, started: true);
    Exercise exercise = Exercise(
        name: 'Squat',
        minRepetitions: 5,
        maxRepetitions: 9,
        executions: [execution, execution2, execution3],
        weigth: 35);
    Workout workout = Workout(
        day: DateTime.now(),
        exercises: [exercise],
        completed: true,
        dayCount: 1);

    // Act
    _dataProvider.saveWorkout(workout);
    final exercises = await _dataProvider.getExercises();
    final sets = await _dataProvider.getSets();

    // Assert
    expect(exercises.isNotEmpty, true);
    expect(exercises[0].name, exercise.name);
    expect(exercises[0].weight, exercise.weigth);

    expect(sets.length, 3);
    expect(sets[0].repetitions, 9);
    expect(sets[1].repetitions, 8);
    expect(sets[2].repetitions, 7);

    expect(sets.length, 3);
    expect(sets[0].repetitions, 9);
    expect(sets[1].repetitions, 8);
    expect(sets[2].repetitions, 7);
  });

  test('successfully convert db data to domain data', () async {
    // Arrange
    Execution execution = Execution(currentRepetitons: 9, started: true);
    Execution execution2 = Execution(currentRepetitons: 8, started: true);
    Execution execution3 = Execution(currentRepetitons: 7, started: true);
    Exercise exercise = Exercise(
        name: 'Squat',
        minRepetitions: 5,
        maxRepetitions: 9,
        executions: [execution, execution2, execution3],
        weigth: 35);
    Workout workoutInDb = Workout(
        day: DateTime.now(),
        exercises: [exercise],
        completed: true,
        dayCount: 1);

    _dataProvider.saveWorkout(workoutInDb);

    // Act

    final workout = await _dataProvider.getWorkoutByDate(workoutInDb.day);

    // Assert
    await expectLater(workout.day.day, workoutInDb.day.day);
  });
```

The only workaround which fixed the `Bad state` issue was to remove the `tearDown`-function. But then i get the

`It looks like you've created the database class MyDatabase multiple times` 

warning.

I'am doing something wrong? Or is the official documentation outdated?

**Comment by simolus3** [comment, marked as answer]
Opening the database in `setUp` and closing it in `tearDown` should work. I'm using the same approach for tests in this package, and it's working fine:

https://github.com/simolus3/moor/blob/a16b34c8d06c919dcd175a399c11b8c48da34bce/moor/test/integration_tests/moor_files_integration_test.dart#L10-L18

My best guess is that you're somehow still using the old database in the second test? Are you storing the database in a top-level or static field somehow? Your test code looks reasonable though.

But also, shouldn't `saveWorkout` return a future and be awaited in your test? If you're not awaiting everything, it may be that the test completes before `saveWorkout` is done. In that case the database may be closed prematurely.

**Reply by rubiktubik** [reply]
awaiting `saveWorkout` did solve the problem! How could I have missed it 🤦‍♂️😅

Thanks for your time!
It's always a please working with moor 👌

**Reply by simolus3** [reply]
Thanks for the kind words :) Btw, you can catch this easier by enabling [unawaited_futures](https://dart-lang.github.io/linter/lints/unawaited_futures.html).
